### PR TITLE
Add args to keybinding

### PR DESCRIPTION
**Pull request by nehayward**
Fix for issue #31372

### CODE REVIEW

File: src/vs/workbench/services/keybinding/electron-browser/keybindingService.ts [hunk 1]
```diff
@@ -160,6 +160,7 @@ interface ContributedKeyBinding {
 	mac?: string;
 	linux?: string;
 	win?: string;
+	args?: [string];
```
Comment (jrieken) [inline src/vs/workbench/services/keybinding/electron-browser/keybindingService.ts]: `[string]` is a funky type. why not `any`?

File: src/vs/workbench/services/keybinding/electron-browser/keybindingService.ts [hunk 2]
```diff
@@ -200,7 +201,7 @@ function isValidContributedKeyBinding(keyBinding: ContributedKeyBinding, rejects
 
 let keybindingType: IJSONSchema = {
 	type: 'object',
-	default: { command: '', key: '' },
+	default: { command: '', key: '', args: [''] },
```
Comment (jrieken) [inline src/vs/workbench/services/keybinding/electron-browser/keybindingService.ts]: not sure about the default value for args... i'd this is a special use-case and maybe doesn't deserve a default

File: src/vs/workbench/services/keybinding/electron-browser/keybindingService.ts [hunk 3]
```diff
@@ -222,6 +223,9 @@ let keybindingType: IJSONSchema = {
 			description: nls.localize('vscode.extension.contributes.keybindings.win', 'Windows specific key or key sequence.'),
 			type: 'string'
 		},
+		args: {
+			'description': nls.localize('scode.extension.contributes.keybindings.args', "Arguments to pass to the command to execute.")
```
Comment (jrieken) [inline src/vs/workbench/services/keybinding/electron-browser/keybindingService.ts]: add a type

File: src/vs/platform/keybinding/common/keybindingsRegistry.ts [hunk 4]
```diff
@@ -46,6 +46,7 @@ export interface IKeybindingRule2 {
 	win?: { primary: Keybinding; };
 	linux?: { primary: Keybinding; };
 	mac?: { primary: Keybinding; };
+	args?: [string];
```
Comment (jrieken) [inline src/vs/platform/keybinding/common/keybindingsRegistry.ts]: args?: any;

File: src/vs/platform/keybinding/common/keybindingsRegistry.ts [hunk 5]
```diff
@@ -190,7 +191,7 @@ class KeybindingsRegistryImpl implements IKeybindingsRegistry {
 		}
 	}
 
-	private _registerDefaultKeybinding(keybinding: Keybinding, commandId: string, weight1: number, weight2: number, when: ContextKeyExpr, source: KeybindingRuleSource): void {
+	private _registerDefaultKeybinding(keybinding: Keybinding, args: [string], commandId: string, weight1: number, weight2: number, when: ContextKeyExpr, source: KeybindingRuleSource): void {
```
Comment (jrieken) [inline src/vs/platform/keybinding/common/keybindingsRegistry.ts]: maybe a different order? `...commandId: string, commandArgs:string, ...`

**Comment by nehayward** [comment]
Thanks @jrieken!

File: src/vs/platform/keybinding/common/keybindingsRegistry.ts [hunk 6]
```diff
@@ -190,7 +191,7 @@ class KeybindingsRegistryImpl implements IKeybindingsRegistry {
 		}
 	}
 
-	private _registerDefaultKeybinding(keybinding: Keybinding, commandId: string, weight1: number, weight2: number, when: ContextKeyExpr, source: KeybindingRuleSource): void {
+	private _registerDefaultKeybinding(keybinding: Keybinding, commandId: string, commandArgs: string, weight1: number, weight2: number, when: ContextKeyExpr, source: KeybindingRuleSource): void {
```
Comment (jrieken) [inline src/vs/platform/keybinding/common/keybindingsRegistry.ts]: wrong type for `commandArgs`

File: src/vs/platform/keybinding/common/keybindingsRegistry.ts [hunk 7]
```diff
@@ -132,13 +133,13 @@ class KeybindingsRegistryImpl implements IKeybindingsRegistry {
 		let actualKb = KeybindingsRegistryImpl.bindToCurrentPlatform(rule);
 
 		if (actualKb && actualKb.primary) {
-			this._registerDefaultKeybinding(createKeybinding(actualKb.primary, OS), rule.id, rule.weight, 0, rule.when, source);
+			this._registerDefaultKeybinding(createKeybinding(actualKb.primary, OS), undefined, rule.id, rule.weight, 0, rule.when, source);
```
Comment (jrieken) [inline src/vs/platform/keybinding/common/keybindingsRegistry.ts]: wrong args order wrt command args and command id

File: src/vs/platform/keybinding/common/keybindingsRegistry.ts [hunk 8]
```diff
@@ -132,13 +133,13 @@ class KeybindingsRegistryImpl implements IKeybindingsRegistry {
...
 		}
 
 		if (actualKb && Array.isArray(actualKb.secondary)) {
 			for (let i = 0, len = actualKb.secondary.length; i < len; i++) {
 				const k = actualKb.secondary[i];
-				this._registerDefaultKeybinding(createKeybinding(k, OS), rule.id, rule.weight, -i - 1, rule.when, source);
+				this._registerDefaultKeybinding(createKeybinding(k, OS), undefined, rule.id, rule.weight, -i - 1, rule.when, source);
```
Comment (jrieken) [inline src/vs/platform/keybinding/common/keybindingsRegistry.ts]: same

File: src/vs/platform/keybinding/common/keybindingsRegistry.ts [hunk 9]
```diff
@@ -147,7 +148,7 @@ class KeybindingsRegistryImpl implements IKeybindingsRegistry {
 		let actualKb = KeybindingsRegistryImpl.bindToCurrentPlatform2(rule);
 
 		if (actualKb && actualKb.primary) {
-			this._registerDefaultKeybinding(actualKb.primary, rule.id, rule.weight, 0, rule.when, source);
+			this._registerDefaultKeybinding(actualKb.primary, rule.args, rule.id, rule.weight, 0, rule.when, source);
```
Comment (jrieken) [inline src/vs/platform/keybinding/common/keybindingsRegistry.ts]: same

File: src/vs/workbench/services/keybinding/electron-browser/keybindingService.ts [hunk 10]
```diff
@@ -200,7 +201,7 @@ function isValidContributedKeyBinding(keyBinding: ContributedKeyBinding, rejects
 
 let keybindingType: IJSONSchema = {
 	type: 'object',
-	default: { command: '', key: '' },
+	default: { command: '', key: ''},
```
Comment (jrieken) [inline src/vs/workbench/services/keybinding/electron-browser/keybindingService.ts]: fails formatting/hygiene check

**Comment by nehayward** [comment]
Okay, I'm able to run and debug, I'm wondering how to test this feature. Should I write a new test for this or add to an existing suite.

**Comment by jrieken** [comment]
I think there is no existing suite to piggy back on so for me it's OK if you manually test this. The easiest is to write an extension that configures an argument that it then passed to a command (a command it defines)

**Comment by jrieken** [comment]
@nehayward Are you planning to finish this?

**Comment by nehayward** [comment]
@jrieken yes I'm just blocked on the testing part. Sorry about the delay.

**Comment by jrieken** [comment]
Do you need more info apart from this: https://github.com/Microsoft/vscode/pull/59911#issuecomment-430158182

**Comment by nehayward** [comment]
@jrieken Is there a good source to get started on making an extension. I can build and run vscode, just having trouble with making an extension. I'd also like to try and write a new unit test and I'm wondering where that should go.

**Comment by jrieken** [comment]
https://code.visualstudio.com/docs/extensions/example-hello-world

**Comment by nehayward** [comment]
Thanks, Okay I got a hello world. I'm trying to grab the argument from the defined keyboard shortcut.

```son
   "contributes": {
        "commands": [
            {
                "command": "extension.sayHello",
                "title": "Hello World"
            }
        ],
        "keybindings":[
            {
                "command": "extension.sayHello",
                "key": "shift+cmd+i",
                "args": "name"
            }
        ]
    },
```

I'm not sure what function to call here to display the args or at least identify the bug. 🐛 

```js
 let disposable = vscode.commands.registerCommand('extension.sayHello', () => {
        // The code you place here will be executed every time your command is executed

        


        // Display a message box to the user
        vscode.window.showInformationMessage('Hello World!');
    });
    
```

Do I need to update this [function](https://github.com/Microsoft/vscode/blob/72fcd264c552ddd94dfbd0807813ba938d8af8d2/src/vs/workbench/services/keybinding/electron-browser/keybindingService.ts#L483)?

**Comment by jrieken** [comment]
Have something like this: `registerCommand('extension.sayHello', (myArgs) => {`

**Comment by nehayward** [comment]
Okay, yeah I'm blocked, it's still undefined. I'm trying to debug but I'm not seeing anything logging when I run vscode then run the extension and try to attach to host.

**Pull request closed without merging**